### DISCUSSION TITLE
Allow symlinks in non-jekyll builds for GitHub Pages deploy action

### DISCUSSION THREAD

**Discussion by scaramallion** [community/community#9104] [Other Feature Feedback, Questions, & Ideas]
I have a couple of projects that have the documentation built by sphinx and then [pushed to X.Y version subdirectories](https://github.com/pydicom/pynetdicom/tree/gh-pages/), along with a symlink called `stable` pointing to the sudirectory of the current release:
```
1.0/
1.1/
1.2/
dev/
stable  # symlink to 1.2/
.nojekyll
```
Since the change to deploying Pages in the "pages build and deployment" action, the `stable` symlink is causing a deploy failure "Error: Artifact could not be deployed. Please ensure the content does not contain any hard links, symlinks and total size is less than 10GB."

Obviously, removing the symlink fixes the deployment error, but having a symlink here is really useful since I don't have to continually update the `stable` directory with the most recent version of the built docs every time I do a point release. It also means I don't have to needlessly duplicate the data.

**Comment by yoannchaudet** [comment]
Sorry for that!

It's actually a regression we are working on.

The new expected behaviors is for symlinks to be dereferenced before they reach the Pages servers. This is not properly happening now which is why you are seing this error.

**Reply by Ordinant** [reply]
My company's entire public doc set is down because of this error. Please fix this ASAP. This is not a _yawn, it's a holiday_ thing.

**Comment by mosabua** [comment]
I love the visibility I get from the GHA usage .. but timing on introducing that on Friday and changing so much that it fails numerous sites seems a bit reckless. If I can somehow help .. ping me

**Comment by mosabua** [comment]
Works now again! Thanks GitHub team!

**Comment by yoannchaudet** [comment]
The fix landed in production a few hours ago.

**Reply by scaramallion** [reply]
Awesome, thanks!

**Comment by jd-lara** [comment]
@yoannchaudet we are still experiencing the same issue https://github.com/NREL-SIIP/PowerSystems.jl/runs/4601044253?check_suite_focus=true

**Reply by yoannchaudet** [reply]
Push a new commit to your Pages site, even if it is empty. This will give you a new workflow.

Re-running a previous workflow will not work here because the issue is in the workflow definition itself.

**Comment by EiffL** [comment]
Not  exactly the same problem, but I'm running into a very similar issue. As far as I can see, the current github action for generating pages contains:
```
touch artifact.tar && tar --dereference --hard-dereference -cvf artifact.tar --exclude=.git --exclude=artifact.tar .
```
which removes symlinks in the tar archive, and in my case where the same heavy folder is symlinked many times, I end up with a massive archive, full of copies of the same files, that fails to upload (error [here](https://github.com/EiffL/talks/actions/runs/2191938825)). 

My repo hasn't been able to publish new versions of my pages since December or something.

Can't symlinks be preserved in the archive?  Is there a deep reason behind the `--dereference` option?

**Reply by yoannchaudet** [reply]
Hello!

Security is the main reason. Symlinks allow for many malicious use cases and have been the target of multiple security bounties we had to patch over the years.

When opening up our distributed file system so Pages deployments could natively happen from Actions, we made the conscientious choice to forbid symlinks altogether.

Another reason is implementation detail. Our current distributed file system supports symlinks. We have plans to relocated it in the cloud where symlinks will not be an option.

Few solutions for you today:

- If you use Jekyll, you can use client side redirections so you don't need to duplicate your content
- If you use something else that does not support client side redirections, you can try to break down your Pages site into multiple sites

In the longer term, we will support custom HTTP headers and very much have support for server side redirections.

**Reply by EiffL** [reply]
Thanks so much :-) That makes a  lot of sense, and I'll look into redirection instead of symlink.

**Comment by rungrotArunon** [comment]
Thanks you bro

**Comment by niedzielski** [comment]
I think this issue is back:

- Broken March 5th v1.1.2: [page build](https://github.com/oidoid/superpatience.com/runs/11775482780), [commit](https://github.com/oidoid/superpatience.com/commit/24c4c70dcf42729c8f0392c652a0233cffa53da7)
- Working February 4th v1.1.1: [page build](https://github.com/oidoid/superpatience.com/runs/11106750428), [commit](https://github.com/oidoid/superpatience.com/commit/8002db225cb90dd9cacd6f3222467ea17d4586b2)

The error is `Site contained a symlink that should be dereferenced: /atlas.json.` but I haven't changed much and nothing to do with the symlink structure AFAIK.

**Reply by niedzielski** [reply]
Yeah, seems to be a regression. [This is a revert of the March 5th v1.1.2 commit back to February 4th v1.1.1](https://github.com/oidoid/superpatience.com/commit/7e9f7640df019c6327dad1db2d5b96db2d89a14e), which worked before but now doesn't.

**Reply by yoannchaudet** [reply]
We stopped allowing symbolic links in non-Actions builds on February 28th (see [changelog](https://github.blog/changelog/2023-02-21-github-pages-deprecating-symlinks-in-non-actions-builds/)).

In your case to fix the behavior you either need to remove the symbolic link from your repository or enable Actions in your repository.

**Reply by niedzielski** [reply]
Works for me. Thank you, @yoannchaudet!

**Reply by FrancisMurtagh-arm** [reply]
> Actions

Hi @yoannchaudet, even with all actions allowed the error still appears

**Reply by yoannchaudet** [reply]
If you have the name of your repo I can look it up. Org IP allow list is something that also makes the build fallback to the legacy build infra.

**Reply by FrancisMurtagh-arm** [reply]
Hi @yoannchaudet, 

This is the repo:
https://github.com/ARM-software/armnn

This is a link to the error:
https://github.com/ARM-software/armnn/runs/11906108965

Thanks for the help.

**Reply by yoannchaudet** [reply]
Oh I get it. You are on a legacy billing plan (Gold). Actions is unavailable on this plan so unfortunately you are currently stuck on our legacy build infrastructure and that means you cannot have symlinks in your repo.

**Reply by nikraj01** [reply]
@yoannchaudet So does that mean we need to upgrade our github plan? What would be the steps required for upgrading?
Thanks

**Reply by chrispat** [reply]
You should be able to do this in the billing section in your organization admin settings.

**Comment by jarrodmillman** [comment]
We are also seeing it on the NX project: https://github.com/networkx/documentation/runs/11718730700


The problem started March 2, 2023: https://github.com/networkx/documentation/commits/gh-pages

**Comment by mmguero** [comment]
I'm seeing this as well, as of a few weeks ago. I had one PNG image file referenced in the markdown used to generate my pages that was giving this error, so I reorganized the directory so I wouldn't need the symlink. But the error just moved down the line to the next symlink in my repo: apparently having any symlinks at all in the repository is causing this error for me.
